commit 3ba06c6fbd651ed3377e584026d1c112b492cc8b
Author: Jouni Malinen <j@w1.fi>
Date:   Fri Aug 27 22:21:13 2010 +0300

    mac80211: Fix signal strength average initialization for CQM events
    
    The ave_beacon_signal value uses 1/16 dB unit and as such, must be
    initialized with the signal level of the first Beacon frame multiplied
    by 16. This fixes an issue where the initial CQM events are reported
    incorrectly with a burst of events while the running average
    approaches the correct value after the incorrect initialization. This
    could cause user space -based roaming decision process to get quite
    confused at the moment when we would like to go through authentication
    and DHCP.
    
    Cc: stable@kernel.org
    Signed-off-by: Jouni Malinen <j@w1.fi>
    Signed-off-by: John W. Linville <linville@tuxdriver.com>

diff --git a/net/mac80211/mlme.c b/net/mac80211/mlme.c
index c8694478cde2..7915726d791e 100644
--- a/net/mac80211/mlme.c
+++ b/net/mac80211/mlme.c
@@ -1553,7 +1553,7 @@ static void ieee80211_rx_mgmt_beacon(struct ieee80211_sub_if_data *sdata,
 	ifmgd->last_beacon_signal = rx_status->signal;
 	if (ifmgd->flags & IEEE80211_STA_RESET_SIGNAL_AVE) {
 		ifmgd->flags &= ~IEEE80211_STA_RESET_SIGNAL_AVE;
-		ifmgd->ave_beacon_signal = rx_status->signal;
+		ifmgd->ave_beacon_signal = rx_status->signal * 16;
 		ifmgd->last_cqm_event_signal = 0;
 	} else {
 		ifmgd->ave_beacon_signal =